### PR TITLE
feat(macos): basic launchctl service support (WIP)

### PR DESCRIPTION
**Pull request by noelmcloughlin**
### PR progress checklist (to be filled in by reviewers)

- [x] Changes to documentation are appropriate (or tick if not required)
- [x] Changes to tests are appropriate (or tick if not required)
- [ ] Reviews completed

---

#### Primary type

- [ ] `[build]`    Changes related to the build system
- [ ] `[chore]`    Changes to the build process or auxiliary tools and libraries such as documentation generation
- [ ] `[ci]`       Changes to the continuous integration configuration
- [x] `[feat]`     A new feature
- [x] `[fix]`      A bug fix
- [ ] `[perf]`     A code change that improves performance
- [ ] `[refactor]` A code change that neither fixes a bug nor adds a feature
- [ ] `[revert]`   A change used to revert a previous commit
- [ ] `[style]`    Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)

#### Secondary type

- [ ] `[docs]`     Documentation changes
- [x] `[test]`     Adding missing or correcting existing tests

### Does this PR introduce a `BREAKING CHANGE`?
No.

### Describe the changes you're proposing

I want to fix errors on MacOS by adding basic support for launchctl.  I personally am unsure I actually need daemon support, but `service.running` generates errors and I don't want failures.  In my implementation the macos daemons are loaded/unloaded to confirm they work.  In a later PR someone may want to change this behaviour.

My implementation follows this gist: https://gist.github.com/ryanmaclean/7792b2d21aee9269c561393c26e0cd7a

**Which fixes this error:**
```
          ID: salt-master
    Function: service.running
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/local/Cellar/salt/2019.2.2/libexec/lib/python3.7/site-packages/salt/state.py", line 1933, in call
                  **cdata['kwargs'])
                File "/usr/local/Cellar/salt/2019.2.2/libexec/lib/python3.7/site-packages/salt/loader.py", line 1951, in wrapper
                  return f(*args, **kwargs)
                File "/usr/local/Cellar/salt/2019.2.2/libexec/lib/python3.7/site-packages/salt/states/service.py", line 928, in mod_watch
                  __salt__['service.stop'](name)
                File "/usr/local/Cellar/salt/2019.2.2/libexec/lib/python3.7/site-packages/salt/modules/mac_service.py", line 439, in stop
                  domain_target, path = _get_domain_target(name)
                File "/usr/local/Cellar/salt/2019.2.2/libexec/lib/python3.7/site-packages/salt/modules/mac_service.py", line 208, in _get_domain_target
                  service = _get_service(name)
                File "/usr/local/Cellar/salt/2019.2.2/libexec/lib/python3.7/site-packages/salt/modules/mac_service.py", line 142, in _get_service
                  raise CommandExecutionError('Service not found: {0}'.format(name))
              salt.exceptions.CommandExecutionError: Service not found: salt-master
     Started: 16:39:33.991196
    Duration: 442.383 ms
     Changes:
```

### Pillar / config required to test the proposed changes

N/A

### Debug log showing how the proposed changes work

```
[ERROR   ] {'salt-master': '4911'}
[ERROR   ] {'com.saltstack.salt.minion': '5372'}
local:
----------
          ID: salt-pkgrepo-install-saltstack-norepo-macos
    Function: test.show_notification
      Result: True
     Comment: No official repository for MacOS
     Started: 11:10:15.928880
    Duration: 0.647 ms
     Changes:   
----------
          ID: /srv/salt/namespaces/saltstack-formulas
    Function: file.directory
      Result: True
     Comment: The directory /srv/salt/namespaces/saltstack-formulas is in the correct state
     Started: 11:10:15.934892
    Duration: 14.513 ms
     Changes:   
----------
          ID: /srv/salt/namespaces/saltstack-formulas/salt-formula_base
    Function: git.latest
        Name: https://github.com/saltstack-formulas/salt-formula.git
      Result: True
     Comment: unless condition is true
     Started: 11:10:15.970065
    Duration: 19.261 ms
     Changes:        
----------
          ID: salt-master-macos
    Function: file.managed
        Name: /Library/LaunchDaemons/com.saltstack.salt.master.plist
      Result: True
     Comment: File /Library/LaunchDaemons/com.saltstack.salt.master.plist is in the correct state
     Started: 11:10:16.670323
    Duration: 5947.426 ms
     Changes:   
----------
          ID: salt-master-macos
    Function: cmd.run
        Name: launchctl load -w /Library/LaunchDaemons/com.saltstack.salt.master.plist
      Result: True
     Comment: Command "launchctl load -w /Library/LaunchDaemons/com.saltstack.salt.master.plist" run
     Started: 11:10:22.618341
    Duration: 16.387 ms
     Changes:   
              ----------
              pid:
                  4894
              retcode:
                  0
              stderr:
                  /Library/LaunchDaemons/com.saltstack.salt.master.plist: service already loaded
              stdout:
----------
          ID: salt-master-macos
    Function: cmd.run
        Name: launchctl unload /Library/LaunchDaemons/com.saltstack.salt.master.plist
      Result: True
     Comment: Command "launchctl unload /Library/LaunchDaemons/com.saltstack.salt.master.plist" run
     Started: 11:10:22.635922
    Duration: 5022.862 ms
     Changes:   
              ----------
              pid:
                  4895
              retcode:
                  0
              stderr:
              stdout:
----------
          ID: salt-master
    Function: file.recurse
        Name: /private/etc/salt/master.d
      Result: True
     Comment: The directory /private/etc/salt/master.d is in the correct state
     Started: 11:10:27.659178
    Duration: 339.176 ms
     Changes:   
----------
          ID: remove-old-master-conf-file
    Function: file.absent
        Name: /private/etc/salt/master.d/_defaults.conf
      Result: True
     Comment: File /private/etc/salt/master.d/_defaults.conf is not present
     Started: 11:10:27.999995
    Duration: 0.317 ms
     Changes:   
----------
          ID: salt-master
    Function: service.running
      Result: True
     Comment: Service salt-master is already enabled, and is dead
     Started: 11:10:28.001031
    Duration: 58.754 ms
     Changes:   
              ----------
              salt-master:
                  4911
----------
          ID: ensure-roster-config
    Function: file.managed
        Name: /private/etc/salt/roster
      Result: True
     Comment: File /private/etc/salt/roster is in the correct state
     Started: 11:10:28.060160
    Duration: 10.325 ms
     Changes:   
----------
          ID: salt-minion-standalone
    Function: file.recurse
        Name: /private/etc/salt/minion.d
      Result: True
     Comment: Recursively updated /private/etc/salt/minion.d
     Started: 11:10:28.070614
    Duration: 285.602 ms
     Changes:   
              ----------
              /private/etc/salt/minion.d/f_defaults.conf:
                  ----------
                  diff:
                      --- 
                      +++ 
                      @@ -534,7 +534,7 @@
                       # files, but can be directed to look at the local file directory setting
                       # defined below by setting it to "local". Setting a local file_client runs the
                       # minion in masterless mode.
                      -#file_client: "remote"
                      +file_client: local
                       
                       # The file directory works on environments passed to the minion, each environment
                       # can have multiple root directories, the subdirectories in the multiple file
----------
          ID: salt-minion-standalone
    Function: service.dead
        Name: com.saltstack.salt.minion
      Result: True
     Comment: Service com.saltstack.salt.minion has been disabled, and is dead
     Started: 11:10:28.360638
    Duration: 1724.613 ms
     Changes:   
              ----------
              com.saltstack.salt.minion:
                  True
----------
          ID: remove-old-standalone-conf-file
    Function: file.absent
        Name: /private/etc/salt/minion.d/_defaults.conf
      Result: True
     Comment: File /private/etc/salt/minion.d/_defaults.conf is not present
     Started: 11:10:30.087878
    Duration: 0.502 ms
     Changes:   
----------
          ID: salt-minion-macos
    Function: file.managed
        Name: /Library/LaunchDaemons/com.saltstack.salt.minion.plist
      Result: True
     Comment: File /Library/LaunchDaemons/com.saltstack.salt.minion.plist is in the correct state
     Started: 11:10:30.088498
    Duration: 6719.806 ms
     Changes:   
----------
          ID: salt-minion-macos
    Function: cmd.run
        Name: launchctl load -w /Library/LaunchDaemons/com.saltstack.salt.minion.plist
      Result: True
     Comment: Command "launchctl load -w /Library/LaunchDaemons/com.saltstack.salt.minion.plist" run
     Started: 11:10:36.809093
    Duration: 32.99 ms
     Changes:   
              ----------
              pid:
                  5163
              retcode:
                  0
              stderr:
              stdout:
----------
          ID: salt-minion-macos
    Function: cmd.run
        Name: launchctl unload /Library/LaunchDaemons/com.saltstack.salt.minion.plist
      Result: True
     Comment: Command "launchctl unload /Library/LaunchDaemons/com.saltstack.salt.minion.plist" run
     Started: 11:10:36.843219
    Duration: 16.907 ms
     Changes:   
              ----------
              pid:
                  5165
              retcode:
                  0
              stderr:
              stdout:
----------
          ID: salt-minion
    Function: file.recurse
        Name: /private/etc/salt/minion.d
      Result: True
     Comment: Recursively updated /private/etc/salt/minion.d
     Started: 11:10:36.860721
    Duration: 4614.17 ms
     Changes:   
              ----------
              /private/etc/salt/minion.d/f_defaults.conf:
                  ----------
                  diff:
                      --- 
                      +++ 
                      @@ -534,7 +534,7 @@
                       # files, but can be directed to look at the local file directory setting
                       # defined below by setting it to "local". Setting a local file_client runs the
                       # minion in masterless mode.
                      -file_client: local
                      +#file_client: "remote"
                       
                       # The file directory works on environments passed to the minion, each environment
                       # can have multiple root directories, the subdirectories in the multiple file
----------
          ID: salt-minion
    Function: service.running
        Name: com.saltstack.salt.minion
      Result: True
     Comment: Service com.saltstack.salt.minion is already enabled, and is dead
     Started: 11:10:41.479774
    Duration: 260.514 ms
     Changes:   
              ----------
              com.saltstack.salt.minion:
                  5372
----------
          ID: remove-old-minion-conf-file
    Function: file.absent
        Name: /private/etc/salt/minion.d/_defaults.conf
      Result: True
     Comment: File /private/etc/salt/minion.d/_defaults.conf is not present
     Started: 11:10:41.745198
    Duration: 0.503 ms
     Changes:   
----------
          ID: salt-minion
    Function: cmd.run
        Name: salt-call --local service.restart com.saltstack.salt.minion --out-file /dev/null
      Result: True
     Comment: Command "salt-call --local service.restart com.saltstack.salt.minion --out-file /dev/null" run
     Started: 11:10:41.746377
    Duration: 8.102 ms
     Changes:   
              ----------
              pid:
                  5376
              retcode:
                  None
              stderr:
              stdout:
----------
          ID: remove-macpackage-salt
    Function: file.absent
        Name: /tmp/salt.pkg
      Result: True
     Comment: File /tmp/salt.pkg is not present
     Started: 11:10:41.754836
    Duration: 0.62 ms
     Changes:   

Summary for macky
-------------
Succeeded: 21 (changed=9)
Failed:     0
-------------

(base) macky:salt bob$ launchctl load /Library/LaunchDaemons/com.saltstack.salt.minion.plist
(base) OCIR-LM20540:salt bob$ launchctl list | grep minion
6949	0	com.saltstack.salt.minion

(base) macky:salt bob$ launchctl load /Library/LaunchDaemons/com.saltstack.salt.master.plist
(base) OCIR-LM20540:salt bob$ launchctl list | grep master
-	13	**salt-master

```

### Documentation checklist

- [x] Updated the `README` (e.g. `Available states`).
- [ ] Updated `pillar.example`.

### Testing checklist

- [x] Included in Kitchen (i.e. under `state_top`).
- [x] Covered by new/existing tests (e.g. InSpec, Serverspec, etc.).
- [x] Updated the relevant test pillar.

### CODE REVIEW

File: salt/minion.sls [hunk 1]
```diff
@@ -55,14 +54,16 @@ salt-minion-macos:
...
+{%- endif %}
 
 salt-minion:
 {% if salt_settings.install_packages %}
   {%- if grains.os == 'MacOS' and salt_settings.salt_minion_pkg_source %}
   macpackage.installed:
     - name: '/tmp/salt.pkg'
     - target: /
-    {# macpackage.installed behaves weirdly with version_check; version_check detects difference but fails to actually complete install. #}
+    {# macpackage.installed behaves weirdly with version_check, decects diff but fails to complete install. #}
```
Comment (daks) [inline salt/minion.sls]: s/decects/detects/ maybe?

Comment (noelmcloughlin) [reply salt/minion.sls]: Thanks, I'm trying to see if CI will work (commits are gone a bit funny). But anyway will see ...

**Comment by noelmcloughlin** [comment]
Added commit to fix unrelated Travis CI issues

**Comment by weswhet** [comment]
Should we also try and fix the service functions in the core code instead of just `cmd.run`ing the launchctl services?

**Comment by noelmcloughlin** [comment]
> Should we also try and fix the service functions in the core code instead of just `cmd.run`ing the launchctl services?

Yes (I think). Sometimes I presume things are by design not bugs ;-|
Is there an open issue on salt?

**Comment by noelmcloughlin** [comment]
@weswhet @daks Is this PR okay?

**Comment by weswhet** [comment]
@noelmcloughlin so the reason `service.running` with `salt-master` is failing is because the `Label` key of the salt-master LaunchDaemon isn't `salt-master` it's `com.saltstack.salt.master`

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>Label</key>
    <string>com.saltstack.salt.master</string>
    <key>RunAtLoad</key>
    <true/>
    <key>KeepAlive</key>
    <true/>
    <key>ProgramArguments</key>
    <array>
        <string>/opt/salt/bin/start-salt-master.sh</string>
    </array>
    <key>SoftResourceLimits</key>
    <dict>
        <key>NumberOfFiles</key>
        <integer>100000</integer>
    </dict>
    <key>HardResourceLimits</key>
    <dict>
        <key>NumberOfFiles</key>
        <integer>100000</integer>
    </dict>
    <!-- uncomment the lines below to debug launchd issues -->
    <!--
    <key>StandardOutPath</key>
    <string>/tmp/salt-minion.out</string>
    <key>StandardErrorPath</key>
    <string>/tmp/salt-minion.err</string>
    -->
  </dict>
</plist>
```

Can you try changing the name from salt-master to `com.saltstack.salt.master`?

**Comment by noelmcloughlin** [comment]
> Can you try changing the name from salt-master to `com.saltstack.salt.master`?

That changes the error. Is something else missing.
```
$ sudo launchctl unload /Library/LaunchDaemons/com.saltstack.salt.minion.plist
$ sudo rm -fr /Library/LaunchDaemons/com.saltstack.salt.minion.plist
```
...
```
          ID: salt-master
    Function: service.running
        Name: com.saltstack.salt.master
      Result: False
     Comment: Service com.saltstack.salt.master is already enabled, and is dead
     Started: 19:03:14.735099
    Duration: 54.605 ms
     Changes:   
              ----------
              com.saltstack.salt.master:
```

**Comment by weswhet** [comment]
So that means that the master process is exiting/failing after launch. Can you add an `init_delay: 10` and see what happens?

Also, what does `sudo launchctl list com.saltstack.salt.master` look like after you get that error?

**Comment by noelmcloughlin** [comment]
First I tried two runs:
1st ```Comment: Service com.saltstack.salt.master is already enabled, and is dead```
2nd 
```
     Comment: Failed to bootstrap service:
              stdout: 
              stderr: /Library/LaunchDaemons/com.saltstack.salt.master.plist: service already loaded
```
Then I added `init_delay: 10` and ran `$ sudo launchctl unload /Library/LaunchDaemons/com.saltstack.salt.master.plist`

Result:
```
     Comment: Service com.saltstack.salt.master is already enabled, and is dead
              Delayed return for 10 second
    $ sudo launchctl list com.saltstack.salt.master
     Could not find service "com.saltstack.salt.master" in domain for system
```

**Comment by weswhet** [comment]
hmmm, is there anything in the master log after running the formula? 

Did you remove the `cmd.run` pieces in the state?

**Comment by noelmcloughlin** [comment]
Yes, I'm using the original formula and running `salt.master` state.  Just this error:
```
[ERROR   ] {'com.saltstack.salt.master': ''}
```

**Comment by noelmcloughlin** [comment]
Hi @weswhet what should I do with this PR?  As mentioned I'm fixing an error:

```
                File "/usr/local/Cellar/salt/2019.2.2/libexec/lib/python3.7/site-packages/salt/modules/mac_service.py", line 142, in _get_service
                  raise CommandExecutionError('Service not found: {0}'.format(name))
              salt.exceptions.CommandExecutionError: Service not found: salt-master
```

**Comment by weswhet** [comment]
hi @noelmcloughlin I added a PR to have macOS better support service names like on other platforms. https://github.com/saltstack/salt/pull/57646

File: salt/master.sls [hunk 2]
```diff
@@ -2,6 +2,27 @@
 {%- from tplroot ~ "/map.jinja" import salt_settings with context %}
 {%- from tplroot ~ "/libtofs.jinja" import files_switch with context %}
 
+{%- if grains.os == 'MacOS' %}
+salt-master-macos:
+  file.managed:
+    - name: /Library/LaunchDaemons/com.saltstack.salt.master.plist
+    - source: https://raw.githubusercontent.com/saltstack/salt/master/pkg/darwin/com.saltstack.salt.master.plist
```
Comment (weswhet) [inline salt/master.sls:9]: you can use https://raw.githubusercontent.com/saltstack/salt/master/pkg/osx/scripts/com.saltstack.salt.master.plist as that is where we build the macOS pkg from.

File: salt/master.sls [hunk 3]
```diff
@@ -2,6 +2,27 @@
...
+    - source_hash: {{ salt_settings.salt_master_macos_plist_hash }}
+    - retry:
+        attempts: 2
+        until: True
+        interval: 10
+        splay: 10
+  cmd.run:
+    - names:
+      - launchctl load -w /Library/LaunchDaemons/com.saltstack.salt.master.plist
```
Comment (weswhet) [inline salt/master.sls:18]: you can add a `service.running` state here with the name `salt-master` or `com.saltstack.salt.master` and watch the file.managed above for changes and the service state will properly restart the service.

File: salt/minion.sls [hunk 4]
```diff
@@ -7,41 +7,69 @@ include:
...
     - runas: {{ salt_settings.rootuser }}
-   {%- endif %}
-{% endif %}
+            {%- endif %}
+
+salt-minion-macos:
+  file.managed:
+    - onlyif: {{ grains.os == 'MacOS' }}
+    - name: /Library/LaunchDaemons/com.saltstack.salt.minion.plist
+    - source: https://raw.githubusercontent.com/saltstack/salt/master/pkg/darwin/com.saltstack.salt.minion.plist
```
Comment (weswhet) [inline salt/minion.sls:48]: similar URL as above but for the minion plist

File: salt/minion.sls [hunk 5]
```diff
@@ -7,41 +7,69 @@ include:
...
+    - source_hash: {{ salt_settings.salt_minion_macos_plist_hash }}
+    - retry:
+        attempts: 2
+        until: True
+        interval: 10
+        splay: 10
+  cmd.run:
```
Comment (weswhet) [inline salt/minion.sls:55]: I think this is a little more "salty" and should handle the restarting properly. If not please open an issue with core and we can fix it.
```  
service.running:
    - name: com.saltstack.salt.minion
    - enable: true
    - watch:
      - file: salt-minion-macos
```

**Comment by myii** [comment]
@noelmcloughlin Just some info for you from our last working group meeting(s):

* @dafyddj has got Kitchen testing on Windows working via. GitHub Actions, it's already merged in the `openvpn-formula` and there is #471 proposed for this formula.
* He also mentioned that something similar should be possible for MacOS -- it would be very useful for PRs such as this one (and the other formulas with MacOS support).

**Comment by noelmcloughlin** [comment]
> * PRs such as this one (and the other formulas with MacOS support).

It would indeed!

**Comment by noelmcloughlin** [comment]
@weswhet  I have implemented the updates.  The PR is good to merge now.
There are issues with salt `service.running` so I raised https://github.com/saltstack/salt/issues/57664

**Comment by noelmcloughlin** [comment]
The file:

`https://raw.githubusercontent.com/saltstack/salt/master/pkg/osx/scripts/com.saltstack.salt.master.plist` 

references `/opt/salt/bin/start-salt-master.sh` but that does not exist on my filesystem.  Where can I find a good plist file?

**Comment by noelmcloughlin** [comment]
Thanks for your reviews, resolved conflict and merging.

**Comment by ixs** [comment]
This PR is broken I think.
On linux minions I'm suddenly seeing the following error:

```
----------
          ID: salt-minion
    Function: service.running
      Result: False
     Comment: The following requisites were not found:
                                 watch:
                                     file: salt-master-macos
                                     file: salt-master
     Started: 14:24:51.537439
    Duration: 0.018 ms
     Changes:
```

**Pull request merged**